### DISCUSSION TITLE
Single block mode

### DISCUSSION THREAD

**Discussion by dresco** [grblHAL/core#3] [Ideas]
Do you think it would be possible to add support for single block mode to the core?

Appreciate it could also be done from a sender with suitable support. My reason for asking - I would like to toggle single stepping through gcode from my modbus control panel project, feels like it would fit well alongside overrides etc..

It looks like it was an original GRBL feature for a while - [Universal-G-Code-Sender/issues/69](https://github.com/winder/Universal-G-Code-Sender/issues/69)

**Comment by dresco** [comment]
If it helps... The code is quite different now, but I found the original 2012 commit that added single block support;

[https://github.com/chamnit/grbl/commit/303cf59f5246b46d0775b31290c04b5712ff4b2c](https://github.com/chamnit/grbl/commit/303cf59f5246b46d0775b31290c04b5712ff4b2c)

**Comment by terjeio** [comment]
> Do you think it would be possible to add support for single block mode to the core?

Yes - easy to add.

I would like switch inputs (controls signals) for this, optional stop and block delete too. The last two are already implemented and has controls signals defined, I'll add a control signal definition for single block mode too.

The control signal handling should be improved in the core as there is only support for handling input asserted events, not deasserted. This means inputs wired to an on/off switch will only trigger an event when switched on... Currently the only way around that is to use toggle action - I do not like that.

I guess switch (or mode) status should be reported back in the `|Pn` real-time report element?

**Reply by terjeio** [reply]
Update commited, use `$S` to toggle mode. Please report back any problems.

**Reply by dresco** [reply]
Wow, that was quick, thank you! I'll test and let you know.. Cheers!

**Reply by terjeio** [reply]
Well, this request was perfect for testing a set of scripts I am working on for updating the repositories, including submodule dependencies, from the subversion master.
